### ISSUE TITLE
Investigate ontoma's lost mappings for future iterations

### ISSUE THREAD

**Issue by ireneisdoomed** [opentargets/issues#1808]
As a **developer** I want **to identify potential enhancements to the OnToma tool by looking at the diseases we used to map to a valid EFO prior to Ontoma's new version**.

## Background

The internal disease mapping process has been drastically changed for the 21.09 evidence due to the implementation of Ontoma v1. This more robust disease mapping has translated in an overall increase in the number of associations and a decrease of false positives.

A way of continuously increase the value of the tool is to analyse the diseases we have left on the way. 

## Analysis
I have extracted the difference of `diseaseFromSource` labels between 21.06 and 21.09 for the sources we map internally. I have excluded Phenodigm from the analysis since their absences might be explained by a change in the score.

[https://app.zenhub.com/files/143733948/37b3f7dc-3e76-4c2a-be16-b4e703aad94d/download](https://app.zenhub.com/files/143733948/37b3f7dc-3e76-4c2a-be16-b4e703aad94d/download)

This is the information contained in the JSON file:
```
root
 |-- diseaseFromSource: string (nullable = true)
 |-- count(diseaseFromSource): long (nullable = false)
 |-- diseaseFromSourceMappedId: string (nullable = false)
 |-- sourceIDs: array (nullable = false)
 |    |-- element: string (containsNull = false)
 |-- associatedTargets: array (nullable = false) # collection of targetFromSourceIds for that lost diseaseFromSource
 |    |-- element: string (containsNull = false)
```

## Tasks

- [ ] Investigate patterns or potential enhancements in the lost disease labels (`diseaseFromSource`).

E.g. "Non-specific early-onset epileptic encephalopathy" used to be mapped to `MONDO_0018614`, a correct mapping. Can the tool be improved to be able to map such case?
```
 diseaseFromSource         | Non-specific early-onset epileptic encephalopathy
 count(diseaseFromSource)  | 39
 sourceIDs                 | [orphanet]
 diseaseFromSourceMappedId | MONDO_0018614
```

**Comment by tskir** [comment]
Unfortunately the attached file doesn't contain the `diseaseFromSourceMappedId` values, perhaps they were lost during export.

However, I don't strictly speaking need those values; I can simply try to map the source disease labels using the latest OnToma and see if they actually do map to something sensible.

I've processed the JSON file with this command to extract just the source labels and the number of occurrences:
```bash
sed -e 's|count[^"]*|count|g' lost_mappings_2109.json | jq -r '.diseaseFromSource + "|" + (.count|tostring)' | tr '|' '\t' > lost_mappings.tsv
```

I'll take a look at it to see how many of those labels have (adequate) mappings with the latest OnToma.

**Comment by tskir** [comment]
I processed all lost mappings with the newest OnToma (v1.0.2):

```bash
cut -f1 lost_mappings.tsv | ontoma \
  --input-type string \
  --outfile remapped.tsv \
  --cache-dir /tmp/efo_cache \
  --columns query,id_normalised,id_ot_schema,id_full_uri,label
```

I then collated all information in this [spreadhseet](https://docs.google.com/spreadsheets/d/1taP0F7iGhvBa32yz8bB2sEK0bGFgsf9vQsL0nZ7_FPA/edit?usp=sharing), and looked at several statistics and observations.

**Comment by tskir** [comment]
### Percentage of traits rescued by OnToma v1.0.2

|                         | Total | Mapped by OnToma v1.0.2 | Not mapped by OnToma v1.0.2 | Mapped % |
|-------------------------|-------|-------------------------|-----------------------------|----------|
| Distinct trait names    | 3,934 |                   1,856 |                       2,078 |      47% |
| Total trait occurrences | 6,778 |                   4,333 |                       2,445 |      64% |

First of all, even when specific mappings might have changed, we can see that, actually, 64% of trait names (when weighted by occurrence) are rescued by OnToma v1.0.2 in the sense that they get some mappings (not necessarily the ones issued by the previous versions; but we do trust that OnToma v1.0.2 are good quality because we ran the benchmarks way back when we released it).

**Comment by tskir** [comment]
### Causes of mappings not being assigned

I've manually checked all traits with occurrence > 10 (which was the threshold back when I helped reviewing the mappings for EVA/ClinVar), and I can confirm there were no situations where an obviously correct mapping was missing.

Sometimes the EFO doesn't contain a suitable term at all; sometimes there is one which might be correct, but it's not obvious to ZOOMA (it returns them with "Medium" mapping confidence) and in the vast majority of cases it's not even obvious to me—that is, to map those terms correctly would require proper manual curation by an expert.

The reason these terms used to receive some mappings _before_ was that pre-v1.0 OnToma used OLS lookup and blindly trusted its results, taking the first term in the search results as correct. It's true that sometimes OLS search is smart enough to figure out the correct term, but in my benchmark of pre-v1.0 OnToma I found that for every correct term it returns, we see about five other glaringly incorrect mappings. All in all it was a very noisy source of mappings, which is why we decided to axe it.

In conclusion, it seems that OnToma v1.0.2 is behaving as expected, and any future improvements of the mapping process (which, I do believe, are quite important) would need to take place as a part of a separate effort.

**Comment by tskir** [comment]
### Minor changes

Over the course of tackling this issue, I introduced several minor changes to make OnToma easier to use. I'll submit them as a PR to the OnToma repository. Once it is merged, I believe this issue can be closed.
